sophia/src/term/literal/_convert.rs

```diff
@@ -0,0 +1,495 @@
+//! Public exported by parent-module `literal`.
+//!
+
+use super::*;
+use crate::ns::xsd;
+use crate::term::{Iri, Result, StaticTerm, TermData, TermError};
+use crate::triple::stream::{SinkError, SourceError, StreamResult};
+
+/// Provide an IRI to specify the semantic datatype.
+pub trait DataType {
+    /// Static datatype of this Rust-type.
+    fn dt() -> &'static Iri<&'static str>;
+
+    /// Datatype of `self`.
+    ///
+    /// This can be used for `enum`s where each variant may has its own
+    /// datatype.
+    fn dt_of(&self) -> &'static Iri<&'static str> {
+        Self::dt()
+    }
+
+    fn dt_term() -> StaticTerm {
+        Iri::from(Self::dt()).into()
+    }
+
+    fn dt_term_of(&self) -> StaticTerm {
+        Iri::from(Self::dt_of(self)).into()
+    }
+}
+
+/// Representation as an RDF literal.
+pub trait AsLiteral<TD: TermData>: DataType {
```

Comment:
Does this really need to be a separate trait from `Datatype`?

Reply:
While writing this code I had in mind having two flavors of `from_literal` traits. One that only cares about the lexical value which is currently implemented in `FromLiteral` maybe it is better called `FromLexicalValue`. The second checking the right datatype, using the `DataType` trait. However, as you mention below there is a type hierarchy in `xsd`, so I haven't figured out how to handle that best.

Reply:
Yes, the type hierarchy is a problem. Some people might prefer to convert Rust types in the most precise datatype (e.g. to ensure roundtrip), while some others might prefer to export all integer values to `xsd:integer` (for homogeneity)...

If we have to chose, I think it is better to keep the most precise information. It is stilss possible to "project" it later to more general datatypes (e.g. adding such a utility method to the list at #52).